File: sdk/go/spinneret/snapshot.go
package spinneret

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"log/slog"
	"os"
	"path"
	"strings"
	"time"

	"google.golang.org/protobuf/encoding/protojson"
)

const (
	// snapshotFormat is the version of the snapshot file layout (shared with the Python SDK).
	snapshotFormat = 1
	// snapshotTokenNamespaceDir names the namespace directory when no namespace is set.
	snapshotTokenNamespaceDir = "_token_namespace"
	// secretRefMarker marks unresolved secret references in config content.
	secretRefMarker = "${secret:"
	// maxSnapshotBytes bounds the size of a snapshot file that is read back.
	maxSnapshotBytes = 32 << 20
	snapshotDirMode  = 0o700
	snapshotFileMode = 0o600
)

// snapshotDocument is the on-disk representation of one config item.
type snapshotDocument struct {
	Format    int             `json:"format"`
	SavedAt   string          `json:"saved_at"`
	Encrypted bool            `json:"encrypted"`
	Item      json.RawMessage `json:"item,omitempty"`
}

// snapshotStore keeps the latest version of each watched item under
// <root>/<host>/<namespace>/<group>/<key>.json (every component
// percent-encoded), written atomically with mode 0600 in 0700 directories.
// Secret items are never written: their older snapshots are removed instead.
type snapshotStore struct {
	root          string
	dir           string // relative to root: <host>/<namespace>
	treatAsSecret func(*ConfigItem) bool
	logger        *slog.Logger
}

func newSnapshotStore(root, host, namespace string, treatAsSecret func(*ConfigItem) bool, logger *slog.Logger) *snapshotStore {
	if namespace == "" {
		namespace = snapshotTokenNamespaceDir
	}
	return &snapshotStore{
		root:          root,
		dir:           path.Join(snapshotComponent(host), snapshotComponent(namespace)),
		treatAsSecret: treatAsSecret,
		logger:        logger,
	}
}

// snapshotComponent encodes an arbitrary string as one safe path component:
// every byte except ASCII letters, digits and "_.-~" is percent-encoded, a
// leading dot is encoded, and the empty string becomes "%00".
func snapshotComponent(name string) string {
	const hexDigits = "0123456789ABCDEF"
	var b strings.Builder
	for i := 0; i < len(name); i++ {
		c := name[i]
		unreserved := c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' || c >= '0' && c <= '9' ||
			c == '_' || c == '.' || c == '-' || c == '~'
		if unreserved && (c != '.' || i > 0) {
			b.WriteByte(c)
			continue
		}
		b.WriteByte('%')
		b.WriteByte(hexDigits[c>>4])
		b.WriteByte(hexDigits[c&0x0f])
	}
	if b.Len() == 0 {
		return "%00"
	}
	return b.String()
}

// relPath returns the snapshot file of an item relative to the root.
func (s *snapshotStore) relPath(group, key string) string {
	return path.Join(s.dir, snapshotComponent(group), snapshotComponent(key)+".json")
}

// isSecret reports whether an item must not be persisted. A predicate that
// panics marks the item as secret.
func (s *snapshotStore) isSecret(item *ConfigItem) (secret bool) {
	if item.GetHasSecretRefs() || strings.Contains(item.GetContent(), secretRefMarker) {
		return true
	}
	if s.treatAsSecret == nil {
		return false
	}
	defer func() {
		if p := recover(); p != nil {
			s.logger.Error("spinneret TreatAsSecret panicked, treating the item as secret",
				slog.String("group", item.GetGroup()), slog.String("key", item.GetKey()))
			secret = true
		}
	}()
	return s.treatAsSecret(item)
}

func (s *snapshotStore) openRoot() (*os.Root, error) {
	if err := os.MkdirAll(s.root, snapshotDirMode); err != nil {
		return nil, err
	}
	return os.OpenRoot(s.root)
}

// save writes the snapshot of item and reports whether a file was written
// (false for secret items, whose older snapshot is removed).
func (s *snapshotStore) save(item *ConfigItem) (bool, error) {
	root, err := s.openRoot()
	if err != nil {
		return false, err
	}
	defer func() { _ = root.Close() }()
	rel := s.relPath(item.GetGroup(), item.GetKey())
	if s.isSecret(item) {
		if err := root.Remove(rel); err != nil && !errors.Is(err, fs.ErrNotExist) {
			return false, err
		}
		return false, nil
	}
	encoded, err := protojson.MarshalOptions{UseProtoNames: true, EmitUnpopulated: true}.Marshal(item)
	if err != nil {
		return false, err
	}
	data, err := json.Marshal(snapshotDocument{
		Format:  snapshotFormat,
		SavedAt: time.Now().UTC().Format("2006-01-02T15:04:05.000000Z"),
		Item:    encoded,
	})
	if err != nil {
		return false, err
	}
	if err := atomicWrite(root, rel, data); err != nil {
		return false, err
	}
	return true, nil
}

// atomicWrite writes data to a temporary file next to rel, syncs it, renames
// it over rel and syncs the directory.
func atomicWrite(root *os.Root, rel string, data []byte) error {
	dir := path.Dir(rel)
	if err := root.MkdirAll(dir, snapshotDirMode); err != nil {
		return err
	}
	tmp := path.Join(dir, "."+path.Base(rel)+"."+NewReportID()+".tmp")
	f, err := root.OpenFile(tmp, os.O_WRONLY|os.O_CREATE|os.O_EXCL, snapshotFileMode)
	if err != nil {
		return err
	}
	_, werr := f.Write(data)
	if werr == nil {
		werr = f.Sync()
	}
	if cerr := f.Close(); werr == nil {
		werr = cerr
	}
	if werr == nil {
		werr = root.Rename(tmp, rel)
	}
	if werr != nil {
		_ = root.Remove(tmp)
		return werr
	}
	if d, err := root.Open(dir); err == nil {
		_ = d.Sync() // best effort: not every platform can sync directories
		_ = d.Close()
	}
	return nil
}

// load reads the snapshot of an item; (nil, nil) when there is none or it
// must be ignored (encrypted by another SDK, flagged as secret).
func (s *snapshotStore) load(group, key string) (*ConfigItem, error) {
	root, err := s.openRoot()
	if err != nil {
		return nil, err
	}
	defer func() { _ = root.Close() }()
	f, err := root.Open(s.relPath(group, key))
	if errors.Is(err, fs.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	defer func() { _ = f.Close() }()
	raw, err := io.ReadAll(io.LimitReader(f, maxSnapshotBytes+1))
	if err != nil {
		return nil, err
	}
	if len(raw) > maxSnapshotBytes {
		return nil, errors.New("snapshot file too large")
	}
	var doc snapshotDocument
	if err := json.Unmarshal(raw, &doc); err != nil {
		return nil, fmt.Errorf("decode snapshot: %w", err)
	}
	if doc.Format != snapshotFormat {
		return nil, fmt.Errorf("unsupported snapshot format %d", doc.Format)
	}
	if doc.Encrypted || len(doc.Item) == 0 {
		return nil, nil
	}
	item := new(ConfigItem)
	if err := (protojson.UnmarshalOptions{DiscardUnknown: true}).Unmarshal(doc.Item, item); err != nil {
		return nil, fmt.Errorf("decode snapshot item: %w", err)
	}
	if item.GetGroup() != group || item.GetKey() != key {
		return nil, errors.New("snapshot belongs to another item")
	}
	if s.isSecret(item) {
		return nil, nil
	}
	return item, nil
}

File: sdk/python/src/spinneret/_snapshot.py
"""Local on-disk snapshots of config items, written atomically."""

from __future__ import annotations

import contextlib
import json
import logging
import os
import tempfile
from collections.abc import Callable
from pathlib import Path
from typing import Any
from urllib.parse import quote

from pydantic import ValidationError

from ._crypto import SnapshotCipher, SnapshotDecryptError
from .models import ConfigItem, format_timestamp, utc_now

__all__ = ["SNAPSHOT_FORMAT", "SecretPredicate", "SnapshotStore"]

logger = logging.getLogger("spinneret.config")

#: Version of the snapshot file layout.
SNAPSHOT_FORMAT = 1

_DEFAULT_NAMESPACE_DIR = "_token_namespace"
_PRIVATE_DIR_MODE = 0o700

#: Decides whether a config item holds secret material that must not be stored in clear text.
SecretPredicate = Callable[[ConfigItem], bool]


def _component(name: str) -> str:
    """Encode an arbitrary string as a single safe path component."""
    encoded = quote(name, safe="")
    if encoded.startswith("."):
        encoded = "%2E" + encoded[1:]
    return encoded or "%00"


def _make_private_dirs(directory: Path) -> None:
    """Create ``directory`` and its missing parents, each with mode ``0700``.

    ``Path.mkdir(parents=True, mode=...)`` applies the mode to the leaf only, which
    would leave intermediate cache directories (named after hosts, namespaces and
    groups) readable by other users.
    """
    missing: list[Path] = []
    current = directory
    while not current.exists() and current != current.parent:
        missing.append(current)
        current = current.parent
    for part in reversed(missing):
        with contextlib.suppress(FileExistsError):
            part.mkdir(mode=_PRIVATE_DIR_MODE)
    if not directory.is_dir():
        raise NotADirectoryError(f"snapshot directory {directory} is not a directory")


def _atomic_write(path: Path, data: bytes) -> None:
    _make_private_dirs(path.parent)
    fd, tmp_name = tempfile.mkstemp(prefix=f".{path.name}.", suffix=".tmp", dir=path.parent)
    try:
        with os.fdopen(fd, "wb") as handle:
            handle.write(data)
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(tmp_name, path)
    except BaseException:
        with contextlib.suppress(OSError):
            os.unlink(tmp_name)
        raise
    _fsync_directory(path.parent)


def _fsync_directory(directory: Path) -> None:
    if os.name != "posix":  # pragma: no cover - directory fsync is POSIX only
        return
    with contextlib.suppress(OSError):
        fd = os.open(directory, os.O_RDONLY)
        try:
            os.fsync(fd)
        finally:
            os.close(fd)


class SnapshotStore:
    """Stores the latest version of each watched config item under a cache directory.

    Layout: ``<root>/<host>/<namespace>/<group>/<key>.json`` with every
    component percent-encoded. Secret items are never written in clear text:
    they are skipped (and any older snapshot removed) unless ``cache_secrets``
    is enabled, in which case they are encrypted with a key derived from the
    token. An item is secret when the server set
    :attr:`ConfigItem.has_secret_refs` (see :attr:`ConfigItem.references_secrets`)
    or the optional ``treat_as_secret`` predicate returns true (a predicate
    that raises marks the item as secret). The predicate can only add secret
    items; it cannot exempt an item flagged by the server.
    """

    def __init__(
        self,
        root: Path,
        *,
        host: str,
        namespace: str,
        token: str,
        cache_secrets: bool = False,
        treat_as_secret: SecretPredicate | None = None,
    ) -> None:
        """Create a store.

        Raises:
            ConfigurationError: ``cache_secrets`` is set but ``cryptography`` is missing.
        """
        self._dir = Path(root) / _component(host) / _component(namespace or _DEFAULT_NAMESPACE_DIR)
        self._cipher: SnapshotCipher | None = SnapshotCipher(token) if cache_secrets else None
        self._treat_as_secret = treat_as_secret

    @property
    def directory(self) -> Path:
        """Directory holding the snapshots of this host and namespace."""
        return self._dir

    def path_for(self, group: str, key: str) -> Path:
        """Snapshot file of a config item."""
        return self._dir / _component(group) / f"{_component(key)}.json"

    def is_secret(self, item: ConfigItem) -> bool:
        """Whether ``item`` must only be persisted encrypted.

        The server flag is checked first, so ``treat_as_secret`` is not called
        for items that already are secret.
        """
        if item.references_secrets:
            return True
        if self._treat_as_secret is None:
            return False
        try:
            return bool(self._treat_as_secret(item))
        except Exception:
            logger.exception(
                "spinneret treat_as_secret predicate raised for %s/%s; treating as secret",
                item.group,
                item.key,
            )
            return True

    def _aad(self, group: str, key: str) -> bytes:
        return f"{group}\x00{key}".encode()

    def save(self, item: ConfigItem) -> bool:
        """Write a snapshot of ``item``.

        Returns:
            ``True`` when a snapshot was written, ``False`` when the item was
            skipped by the secret caching policy or the write failed.
        """
        path = self.path_for(item.group, item.key)
        document: dict[str, Any] = {
            "format": SNAPSHOT_FORMAT,
            "saved_at": format_timestamp(utc_now()),
        }
        try:
            if self.is_secret(item):
                if self._cipher is None:
                    self._remove(path)
                    return False
                plaintext = item.model_dump_json().encode("utf-8")
                document["encrypted"] = True
                document["envelope"] = self._cipher.encrypt(
                    plaintext, self._aad(item.group, item.key)
                )
            else:
                document["encrypted"] = False
                document["item"] = item.model_dump(mode="json")
            _atomic_write(path, json.dumps(document, ensure_ascii=False).encode("utf-8"))
        except OSError as exc:
            logger.warning(
                "spinneret could not write config snapshot %s/%s: %s", item.group, item.key, exc
            )
            return False
        return True

    def load(self, group: str, key: str) -> ConfigItem | None:
        """Read the snapshot of a config item, or ``None`` when unavailable."""
        path = self.path_for(group, key)
        try:
            raw = path.read_bytes()
        except FileNotFoundError:
            return None
        except OSError as exc:
            logger.warning("spinneret could not read config snapshot %s/%s: %s", group, key, exc)
            return None
        try:
            item = self._decode(raw, group, key)
        except (ValueError, ValidationError, SnapshotDecryptError) as exc:
            logger.warning(
                "spinneret ignored invalid config snapshot %s/%s: %s",
                group,
                key,
                type(exc).__name__,
            )
            return None
        if item is not None and (item.group != group or item.key != key):
            logger.warning("spinneret ignored mismatched config snapshot %s/%s", group, key)
            return None
        return item

    def _decode(self, raw: bytes, group: str, key: str) -> ConfigItem | None:
        document = json.loads(raw.decode("utf-8"))
        if not isinstance(document, dict) or document.get("format") != SNAPSHOT_FORMAT:
            raise ValueError("unsupported snapshot format")
        if document.get("encrypted"):
            if self._cipher is None:
                return None
            envelope = document.get("envelope")
            if not isinstance(envelope, dict):
                raise ValueError("encrypted snapshot without envelope")
            plaintext = self._cipher.decrypt(envelope, self._aad(group, key))
            return ConfigItem.model_validate_json(plaintext)
        return ConfigItem.model_validate(document.get("item"))

    def _remove(self, path: Path) -> None:
        with contextlib.suppress(FileNotFoundError):
            path.unlink()
